### ISSUE TITLE
Slower than Swift?

### ISSUE THREAD

**Issue by proyb6** [dvidelabs/flatcc#5]
https://medium.com/@icex33/10-thousand-times-faster-swift-737b1accd973#.heh3ty70g

**Comment by mikkelfj** [comment]
I haven't looked at the benchmark in Swift, but it is really hard to prevent the optimizer from optimizing out a checksum. The times achieved in C and C++ are both about 30 ns per buffer which checks more than 10 fields including vtable lookup, loop overhead and checksum update. A really fast hash lookup with everything in L1 cache can be done in 5 ns on a top-end macbook pro. I'd be very surprised if Swift can beat a hash lookup when also doing any kind of work.

But if Swift uses a decent optimizer and avoids allocation overhead, there is also no reason for it to be much slower either. So great job optimizing Swift, but be careful with the benchmark.

You will see the decode function taking a sum argument for each iteration to prevent the optimizer from removing code, and even this is not reliable as optimizers improve:

int64_t decode(flatcc_builder_t *B, void *buffer, size_t size, int64_t sum)

https://github.com/dvidelabs/flatcc/blob/master/test/benchmark/benchflatcc/benchflatcc.c

**Comment by mikkelfj** [comment]
Also see the last line in each iteration:
return sum + 2 * sum;

This makes the checksum non-linear without adding much overhead. Without this I couldn't get clang to slow down, and the numbers were through the roof. In C++ the benchmark also does some tricks, but these a simpler and didn't fool the C optimizer.

However, I can see some caching being used in Swift. When iterating over a large vector of table elements, you can probably construct accessors that are faster than a generic api because you don't have to assume invalid vtable offsets due to versioning for each table element visited. But then again, the optimizer likely caches most of this, and after all, it is faster than memory bandwidth already.

Perhaps @mzaks has some input?

**Comment by mzaks** [comment]
Hi Mikkel,

honestly, I have no idea why it become so fast. When we were doing clever things we tipped the 300ms mark. After doing the simplest thing possible without any caching on our side it magically went super fast.

Here is a gist with functions which are the infrastructure for reading:
https://gist.github.com/mzaks/60101c8e51774f23486823b0d3de2c0f

No caching, just byte hopping and type conversions.

And this is the function which is executed in the loop:
```swift
func flatDecodeDirect(buffer : UnsafePointer<UInt8>) -> Int{
    
    let fooBarContainerOffset = getFooBarContainerRootOffset(buffer)
    
    var sum:Int = 1
    
    sum = sum + Int(getLocationFrom(buffer, fooBarContainerOffset: fooBarContainerOffset).count)
    sum = sum + Int(getFrootFrom(buffer, fooBarContainerOffset: fooBarContainerOffset).rawValue)
    sum = sum + (getInitializedFrom(buffer, fooBarContainerOffset: fooBarContainerOffset) ? 1 : 0)
    
    for i in 0..<getListCountFrom(buffer, fooBarContainerOffset: fooBarContainerOffset) {
        let foobarOffset = getFooBarOffsetFrom(buffer, fooBarContainerOffset: fooBarContainerOffset, listIndex: i)
        sum = sum + Int(getNameFrom(buffer, fooBarOffset: foobarOffset).count)
        sum = sum + Int(getPostfixFrom(buffer, fooBarOffset: foobarOffset))
        sum = sum + Int(getRatingFrom(buffer, fooBarOffset: foobarOffset))
        
        let bar = getSiblingFrom(buffer, fooBarOffset: foobarOffset)
        
        sum = sum + Int(bar.ratio)
        sum = sum + Int(bar.size)
        sum = sum + Int(bar.time)
        
        let foo = bar.parent
        sum = sum + Int(foo.count)
        sum = sum + Int(foo.i_d)
        sum = sum + Int(foo.length)
        sum = sum + Int(foo.prefix)
    }
    return sum
}
```
It is almost one to one what your C code and also the C++ code does.
`flatDecodeDirect` function gets a pointer to the buffer and just reads the values out. No caching.
I tried to switch `return sum` to `return sum + 2 * sum` this also seems not to change much. The execution time is around 0.35 ms.

Here is how it is measured:

```swift
for _ in 0..<inner_loop_iterations {
   //...
        let time9 = CFAbsoluteTimeGetCurrent()
        for _ in 0..<iterations {
            let result = flatDecodeDirect(outputData)
            assert(result == 8644311667)
            total = total + UInt64(result)
        }
        let time10 = CFAbsoluteTimeGetCurrent()
   //...
        direct = direct + (time10 - time9)
```

And it stays close to this result, even if we use struct based API.

https://gist.github.com/mzaks/deac08a7ffc4c9cab60ec84a41adad67

I guess under the hood those so called computed properties on structs become functions which also get inlined, but the API is much nicer to use:

```swift
func flatuseStruct(buffer : UnsafePointer<UInt8>) -> Int
{
    var sum:Int = 1
    
    // this struct or copies of it are only valid as long as pointer
    // to the underlying data is valid
    // should use createInstance() for a long-term usable mutable object instance
    // if needed, but the struct interface is good for lazy stream processing
    let foobarcontainer = FooBarContainerStruct(buffer)
    
    sum = sum + Int(foobarcontainer.location.count)
    sum = sum + Int(foobarcontainer.fruit.rawValue)
    sum = sum + (foobarcontainer.initialized ? 1 : 0)
    
    for i in 0..<foobarcontainer.list.count {
        let foobar = foobarcontainer.list[i]
        sum = sum + Int(foobar.name.count)
        sum = sum + Int(foobar.postfix)
        sum = sum + Int(foobar.rating)

        let bar = foobar.sibling
        
        sum = sum + Int(bar.ratio)
        sum = sum + Int(bar.size)
        sum = sum + Int(bar.time)
        
        let foo = bar.parent
        sum = sum + Int(foo.count)
        sum = sum + Int(foo.i_d)
        sum = sum + Int(foo.length)
        sum = sum + Int(foo.prefix)
    }
    return sum
}
```

The test is run with `-Os` Optimisation Level. And it seems that it does brutal amount of magic :).

If you can spot something suspicious please let me know.

**Comment by mikkelfj** [comment]
Can you provide an assembly dump?

You could also try
  return ((sum >> 16) ^ sum) * 0x45d9f3bUL;
in both C and Swift for comparison.
You still risk all the constant factors being moved out of the inner loop, but at least it will be looping. The next step is to use volatile or atomic variables, or compiler barriers, but I am not sure if Swift has a suitable construct. OSMemoryBarrier() is likely too expensive.
You could also try to switch between two identical buffers depending on the the lower bit of the checksum if you use the above hash.

**Comment by mzaks** [comment]
Ok so if I do `((sum >> 16) ^ sum) * 0x45d9f3b`

I get an overflow
```swift
        for _ in 0..<iterations {
            let result = flatDecodeDirect(outputData)
//            assert(result == 8644311667)
            total1 = total1 + UInt64(result) //Crash!!!
        }
```

on addition. Swift is a safe language 😕

However if I switch `total1 = total1 + UInt64(result)` to 
`total1 = UInt64(result)` 
I get 12ms. 
And if I do this `total1 = (total1 << 8) + UInt64(result)` I get 64ms. 😔
Haven't done the assembly dump yet because than I guess I need to isolate the example.

**Comment by mikkelfj** [comment]
64 ms is 64 ns per operation if you do 1 mill ops. The absolute best you can expect is something like 8 additions per nano second on a 2GHz machine with 4 instructions in parallel and no branching. 8 times that isn't bad for the work being done, even if slower than C. Rather, you should be happy you got that result because now you can meaningfully find slow operations as they pop up.

It's great you can do struct fields so cleanly. Don't waste that in the hunt for absolute performance if you can keep it below 100ns/buffer. For Strings I would probably provide an alternative interface i.e. possibly both a fast and slow approach. Aside from performance, it is best if invalid UTF-8 strings can also be handled cleanly.

**Comment by mzaks** [comment]
It seems like my benchmarks were not correct due to 
https://en.wikipedia.org/wiki/Loop-invariant_code_motion

I changed the implementation as following:
https://gist.github.com/mzaks/eeeaaa45f230e45299cc7f044a930edc

And now I get around 42ms.
This implementation does a bit more, it stores additional variable, casts it from UInt to Int, but in total I guess this is a more realistic benchmark, and don't let compiler move 
`let result = flatDecodeDirect(outputData)`
outside of the for loop.

**Comment by mikkelfj** [comment]
As I understand it, Swift uses LLVM as optimizer which does these things regardless of the high level language. Apple has worked on LLVM for years both for clang C and for OpenCL and their graphics stack, so you would expect some decent results, although with Swift being a younger language than C, there will be some rough edges.

Overall I think 42ms @ 42ns/op is an outstanding result if you get a clean API to match, although I can understand the initial disappointment after beating C by a factor 70 :-)

Now, if you can also encode buffers at 500-600ns and reuse allocated space with clear between runs, that would be great.

Note that in flatcc there is also a load test than runs through 1000 monsters. The performance profile is rather different - less API overhead but more memory to visit.

**Comment by mzaks** [comment]
I will have a look, thank you! :)

**Comment by proyb6** [comment]
@mzaks Have you tried to compile unsafe optimization mode and how fast will that be?

**Comment by mzaks** [comment]
@proyb6 you mean `-Ofast`
Looks the same as `-Os`. I might come back to decoding and try to squeeze more out of it, but right now I am quite happy with 42ms for one million decode+use.

**Comment by proyb6** [comment]
Is using -Ounchecked that can often faster than C in some case and you know the code well could be a fit for benchmark.

**Comment by mzaks** [comment]
no luck there 
`-Ounchecked` gives
```
75,38 ms direct
168,97 ms using struct
```
compared to 
```
46,04 ms direct
41,65 ms using struct
```

with `-Os`

**Comment by gwvo** [comment]
@mzaks:
I second what Mikkel says. There's no way the current Swift does anything significantly faster than C/C++, unless it was optimized out. Until you manage to make the benchmark not optimize anything out, you'll have very artificial numbers.

I see you also still haven't corrected your article.. that may not give a good impression to any visitors.

**Comment by mzaks** [comment]
@gwvo 
Hi Wouter,
I wrote an update paragraph to my post, the moment folks on ycombinator forum told me about "Loop Invariant Code Motion" and I changed the test to avoid it.
https://news.ycombinator.com/item?id=11713047

That said, I believe, the title of the blog post and paragraphs I wrote before this realisation, should stay untouched, because I see this post as a personal journey. 
My posts are normally read by some friends and my self. I write them mainly for future me to memories experiences I made. This one had a bit more traction. However, that doesn't change the fact that, it is not a scientific paper. It is my personal "diary". I made a mistake and learned something. This is all part of the story. Changing the middle of a story, because I know how it will ends, ruins the story. Comments like following shows me that it resonates with people in the way that I hope it would:
https://medium.com/@The_Dudi/aaaghh-update-slightly-ruined-it-but-this-is-still-good-knowledge-thank-you-53a58c8a5e7e#.hsy4qgg3t

I don't want to create any misconceptions or mislead people. I just want to show that FlatBuffers and performance tuning is a very exciting topic and would like to encourage the reader to try it out herself.

However I am open to discuss this standpoint. I might be wrong again.
